### PR TITLE
Subprotocol example

### LINKED ISSUE #7130: Add example crate for custom rlpx subprotocol
### Describe the feature

Add a new example that showcases rlpx subprotocols, similar to this simple Ping/Pong protocol

https://github.com/paradigmxyz/reth/blob/c9c269b64f43932fccc45cebd156d0a091e22103/crates/net/network/tests/it/multiplex.rs#L25-L26

but something more sophisticated

A protocol can be installed via:

https://github.com/paradigmxyz/reth/blob/c9c269b64f43932fccc45cebd156d0a091e22103/crates/net/network/src/network.rs#L210-L214

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by loocapro**
Closes https://github.com/paradigmxyz/reth/issues/7130

Ref: https://github.com/paradigmxyz/reth/pull/8113

### CODE REVIEW

File: examples/custom-rlpx-subprotocol/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "custom-rlpx-subprotocol"
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/Cargo.toml:2]: we now prefix all examples with example- so we can easily filter them in CI

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great,

to complete this example we should create separate network instance that also supports the protocol and connect to the launched node

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        // spawn the second network instance
+        let subnetwork = NetworkManager::new(net_cfg).await?;
+        let subnetwork_peer_id = *subnetwork.peer_id();
+        let subnetwork_peer_addr = subnetwork.local_addr();
+        let subnetwork_handle = subnetwork.peers_handle();
+        node.task_executor.spawn(subnetwork);
+
+        // connect the launched node to the subnetwork
+        node.network.peers_handle().add_peer(FixedBytes(*subnetwork_peer_id), subnetwork_peer_addr);
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:50]: I don't think the FixedBytes is required here

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+use reth::builder::NodeHandle;
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:1]: This should briefly document what this example does

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        let (tx, rx) = oneshot::channel();
+        peer0_conn.send(CustomCommand::Message { msg: "hello!".to_owned(), response: tx })?;
+        let response = rx.await?;
+        assert_eq!(response, "hello!");
+
+        // send a ping message from peer1 to peer0
+        let (tx, rx) = oneshot::channel();
+        peer1_conn.send(CustomCommand::Message { msg: "world!".to_owned(), response: tx })?;
+        let response = rx.await?;
+        assert_eq!(response, "world!");
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:83]: we also want to print something to stdout here to demonstrate this is working

File: examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
+use reth_eth_wire::{protocol::Protocol, Capability};
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs:1]: ideally we reference the rlpx spec here:

https://github.com/ethereum/devp2p/blob/master/rlpx.md

and mention that this is a simple ping pong protocol that also supports sending messages

File: examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    pub fn pong() -> Self {
+        Self {
+            message_type: CustomRlpxProtoMessageId::Pong,
+            message: CustomRlpxProtoMessageKind::Pong,
+        }
+    }
+
+    /// Creates a new `CustomRlpxProtoMessage` with the given message ID and payload.
+    pub fn encoded(&self) -> BytesMut {
+        let mut buf = BytesMut::new();
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs:54]:
```suggestion
        let mut buf = Vec::new();
```

File: examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs:1]: let's rename this proto.rs

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome, only have a few doc nits and suggestions

File: examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs (hunk 6 above)
Comment (loocapro) [reply examples/custom-rlpx-subprotocol/src/subprotocol/protocol/message.rs:54]: @mattsse i cant change this as the ConnectionHandler expects the associated type as a Stream of BytesMut:

https://github.com/paradigmxyz/reth/blob/74f5a852365c451217cbfa49d90bb55623c05b23/crates/net/network/src/protocol.rs#L47

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+        // establish connection between peer1 and peer0
+        let peer1_to_peer0 = from_peer1.recv().await.expect("peer1 connecting to peer0");
+        let peer1_conn = match peer1_to_peer0 {
+            ProtocolEvent::Established { direction: _, peer_id: peer1_id, to_connection } => {
+                assert_eq!(peer1_id, *peer_id);
+                to_connection
+            }
+        };
+
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:81]: println that we established a connection

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        // send a ping message from peer0 to peer1
+        let (tx, rx) = oneshot::channel();
+        peer0_conn.send(CustomCommand::Message { msg: "hello!".to_owned(), response: tx })?;
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:84]: println send msg

File: examples/custom-rlpx-subprotocol/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        let response = rx.await?;
+        assert_eq!(response, "hello!");
+
```
Comment (mattsse) [inline examples/custom-rlpx-subprotocol/src/main.rs:87]: println received msg

**Review by mattsse — CHANGES_REQUESTED** [review]
this doesn't quite work yet because the additional network is launched with the same default ports as the node

you can set its ports to 0

**Comment by mattsse** [comment]
same issue with discovery port, we can disable discovery for the additional network.

please make sure this actually runs :)

**Review by mattsse — APPROVED** [review]
> 2024-06-26T15:53:21.713076Z  INFO New message received response="hello!"
2024-06-26T15:53:21.713832Z  INFO New message received response="world!"

great, working as expected

**Pull request merged**